examples/chain-of-density/chain_of_density.py

```diff
@@ -0,0 +1,223 @@
+    entity_ctx: str,
+    error_msgs: List[str] = [],
+    remaining_retries=3,
+):
+    # # We then perform a new summary and validate that the entity density has increased ( We have not lost any entities )
+    completion = openai.ChatCompletion.create(
+        model="gpt-3.5-turbo-16k",
+        functions=[RewrittenSummary.openai_schema],
+        function_call={"name": RewrittenSummary.openai_schema["name"]},
+        max_retries=2,
+        messages=[
+            {
+                "role": "system",
+                "content": "You are about to be given an article, an existing summary of the article and some new entities. Please use the information to rewrite the summary to make it denser and more concise so that it covers every entity and detail from the previous summary plus the missing entities",
+            },
+            {"role": "user", "content": f"Here is the article : {article}"},
+            {
+                "role": "user",
+                "content": f"Here is the most recent article : {existing_summary}",
+            },
+            {
+                "role": "user",
+                "content": f"Here is some information on entities you should include in the rewritten summary: {entity_ctx}",
+            },
+            *error_msgs,
+        ],
+    )
+    try:
+        new_summary = RewrittenSummary.from_response(
+            completion, validation_context={"prev_summary": existing_summary}
+        )
+        return new_summary
+    except (ValidationError, JSONDecodeError) as e:
+        if remaining_retries == 0:
+            raise e
+        error_msgs = []
+        error_msgs.append(dict(**completion.choices[0].message))
+        error_msgs.append(
+            {
+                "role": "user",
+                "content": f"Recall the function correctly, exceptions found\n{e}",
+            }
+        )
+        return rewrite_summary(
+            article,
+            existing_summary,
+            entity_ctx,
+            error_msgs,
+            remaining_retries=remaining_retries - 1,
+        )
+
+
+def summarize_article(article: str, summary_steps: int = 3):
+    summary_chain = []
+    # We first generate an initial summary
+    summary: InitialSummary = openai.ChatCompletion.create(
+        model="gpt-3.5-turbo",
+        response_model=InitialSummary,
+        messages=[
+            {
+                "role": "system",
+                "content": "You will generate an increasingly concise, entity-dense summary of the following Article. ",
+            },
+            {"role": "user", "content": f"Here is the Article: {article}"},
+        ],
+        max_retries=2,
+    )
+    summary_chain.append(summary.summary)
+
+    # # Then we perform a denser summarization
+    for _ in range(summary_steps):
+        # At the end of each step, we compare the most recent summarization and see what additional missing entities we need to perform
+        missing_entities_response = openai.ChatCompletion.create(
```

Comment:
Goal is to use Spacy here right?

Reply:
Found spacy a bit unreliable in terms of entity detection even with the 500mb large model and so I figured I would switch back to a llm validator